Dockerfile

```diff
@@ -1,26 +1,25 @@
-FROM node:8-alpine
+FROM node:12-alpine
 MAINTAINER "Dan Farrelly <daniel.j.farrelly@gmail.com>"
 
 ENV NODE_ENV production
 
 RUN apk add --no-cache curl
 
-RUN mkdir -p /usr/src/app
-WORKDIR /usr/src/app
+USER node
 
-ADD package*.json /usr/src/app/
+WORKDIR /home/node
+
+COPY --chown=node:node . /home/node
 
 RUN npm install && \
     npm prune && \
     npm cache clean --force && \
     rm -rf /tmp/*
 
-ADD . /usr/src/app/
-
-EXPOSE 80 25
+EXPOSE 1080 1025
```

Comment:
I don't get the benefit of not-running the app on standard port?

#244 mention security issue without telling what the issue is.

Reply:
Well, it's general convention to not publishing a dev container to the world. Considering that this is used in dev environments, it should be restricted to a port which isn't by default open. All distro firewalls have port 80 opened by default, otherwise http(s) traffic would be blocked by default. This isn't a problem with expose, because expose is used to link containers with a network created by docker itself. However, considering that the maildev server is running in broadcast mode (0.0.0.0), this would mean that every time that a user wants to view the maildev webpage, the user would publish the container to the world. Having these ports set on to higher port values would allow firewalls to block this by default.

Another point could be made regarding that port sniffers would try the default used ports first. Port 80 in combination with basic firewall settings would be really easily found, considering that port 80 is the default webserving port.

The last point is regarding podman. Currently fedora supports running rootless containers by default, and this is becoming a standard really quickly. Ports below 1024 aren't available to users, without setting a kernel flag. This change would therefore make maildev future proof.

Reply:
your assumption are wrong. You don't need to expose container port to host machine, and if you do it, you don't have to use the same port in your mapping.

This is exactly what is performed with nginx or databases in Dev environment.

That's a huge drawback just for Fedora users (sI really wonder how they run nginx/mysql/...). 

I suggest to watch other distribution (if they restrict root users too) and other Images that expose port <1024 before applying such change.

Reply:
what's about https://medium.com/elbstack/docker-go-and-privileged-ports-d6354db472c3 to use standard port with non root user?.

Reply:
> your assumption are wrong. You don't need to expose container port to host machine, and if you do it, you don't have to use the same port in your mapping.

Sure, you can remap the ports however you like. However a sensible safe default image is better than having to retain to custom configuration. These changes are meant to make sure  that beginners and/or overlooked mistakes don't get you exposed for no reason with default docker conventions. 

> This is exactly what is performed with nginx or databases in Dev environment.

In contrast to containers, nginx doesn't work without configuration. You can't start the nginx service and expose your dev applications to the internet, without actually having to think about how you are exposing it. 

> That's a huge drawback just for Fedora users (sI really wonder how they run nginx/mysql/...).

Well you can just set `net.ipv4.ip_unprivileged_port_start=0` and allow users to use privileged ports. It isn't just in reference to fedora users, k8s and docker can currently run in some form rootless, lxc as well. Podman on fedora is just an example. Running podman rootless on arch myself without any hassle.

Running rootless nginx and mysql containerized isn't any different than running them as root. Also keep in mind that the default mysql port is 3306, therefore working without any kernel flags.

> I suggest to watch other distribution (if they restrict root users too) and other Images that expose port <1024 before applying such change.

I'm not sure what you're trying to say. The PR changes have nothing to do with restricting root users and privileged ports. If you mean that I should check other maildev like apps to get inspired. Well...
- https://github.com/mailhog/MailHog/blob/master/Dockerfile
- https://github.com/sj26/mailcatcher/blob/master/Dockerfile

Reply:
> what's about https://medium.com/elbstack/docker-go-and-privileged-ports-d6354db472c3 to use standard port with non root user?.

Are you seriously trying to teach me how to expose the container to 80/25, just so you can keep using the ports? :rofl:

Reply:
this conversation is meaningless

Talking about using a dev tool in K8s, exposing the container network to the Host machine, ignoring Images that expose 25/80, being arrogant.

Reply:
> Talking about using a dev tool in K8s

I'm not talking about any dev tool?

> exposing the container network to the Host machine

I'm aware of the ambiguity, but I was referring to the act of exposing the container to the world.

> ignoring Images that expose 25/80

What do you mean?

> being arrogant

The last comment was meant to be playful.. But anyway,

The only way I can see port 80 being a security issue is that beginner- or configuration mistakes are made, as I have tried to make clear in my above comments (I have edited them a bit so they are more clear regarding this). Expose is meant as a type of documentation, and the Dockerfile is mostly the first reference of needed ports when executing `docker run --publish x:y ...`. Having this set to a sensible default in combination with running the container as the node user, would make it more secure for beginners and it overlaps perfectly with default rootless container setups.

Reply:
Changing from the root user to a `node` user and using a port >1000 apply the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege). It's a smart best practice to follow and as this is open source software, we should default to the least priviledge. If someone cares to run their own images as root, that's fine, but let's have good defaults for the most broad usage.